sopel/logger.py

```diff
@@ -30,39 +34,109 @@ def formatException(self, exc_info):
+                'propagate': False,
+                'handlers': ['raw'],
+            },
+            # asynchat exception logger
+            'sopel.exceptions': {
+                'level': 'INFO',
+                'propagate': False,
+                'handlers': ['exceptionfile'],
+            },
+        },
+        'handlers': {
+            # output on stderr
+            'console': {
+                'level': 'DEBUG',
+                'class': 'logging.StreamHandler',
+                'formatter': 'sopel',
+            },
+            # generic purpose log file
+            'logfile': {
+                'level': 'DEBUG',
+                'class': 'logging.handlers.TimedRotatingFileHandler',
+                'filename': os.path.join(
+                    log_directory, settings.basename + '.sopel.log'),
+                'when': 'midnight',
+                'formatter': 'sopel',
+            },
+            # catched error log file
+            'errorfile': {
+                'level': 'ERROR',
+                'class': 'logging.handlers.TimedRotatingFileHandler',
+                'filename': os.path.join(
+                    log_directory, settings.basename + '.error.log'),
+                'when': 'midnight',
+                'formatter': 'sopel',
+            },
+            # uncaught error file
+            'exceptionfile': {
+                'level': 'ERROR',
+                'class': 'logging.handlers.TimedRotatingFileHandler',
+                'filename': os.path.join(
+                    log_directory, settings.basename + '.exceptions.log'),
+                'when': 'midnight',
+                'formatter': 'sopel',
+            },
+            # raw IRC log file
+            'raw': {
+                'level': 'DEBUG',
+                'class': 'logging.handlers.TimedRotatingFileHandler',
+                'filename': os.path.join(
+                    log_directory, settings.basename + '.raw.log'),
+                'when': 'midnight',
+                'formatter': 'raw',
+            },
+        },
+    }
+    dictConfig(logging_config)
 
 
+@deprecated(
+    'use logging.getLogger(__name__) instead',
+    version='7.0',
+    removed_in='8.0')
 def get_logger(name=None):
     """Return a logger for a module, if the name is given.
 
-    This is equivalent to `logging.getLogger('sopel.modules.' + name)` when
-    name is given, and `logging.getLogger('sopel')` when it is not. The latter
-    case is intended for use in Sopel's core; modules should call
-    `get_logger(__name__)` to get a logger."""
+    This is equivalent to ``logging.getLogger('sopel.modules.' + name)`` when
+    name is given, and ``logging.getLogger('sopel')`` when it is not.
+    The latter case is intended for use in Sopel's core; modules should call
+    ``get_logger(__name__)`` to get a logger.
+
+    .. deprecated:: 7.0
+
+        Use ``logging.getLogger(__name__)`` instead, or
+        ``logging.getLogger('sopel.plugins.YOURPLUGINNAME')``.
```

Comment:
What are the possible consequences if we recommend simply calling `logging.getLogger('YOURPLUGINNAME')`?

We probably don't want to encourage plugin developers to insert their log entries into an existing namespace. (This also ties into that IRC discussion we started the other day about the mess we now have with `sopel.plugins` being used for the machinery when it should ideally be the new name for `sopel.modules`…)

I like that, even though it's being deprecated, you still made an effort to fix the docstring to render inline-code properly in Sphinx. 😁

Reply:
Say that `__name__` is `sopel_modules.myplugin`, then it won't pollute the namespace. Same for `sopel_myplugin`.

I'm still thinking about tools to make plugin logging configuration easier, though.

Reply:
Also I realized that I misunderstood your comment.

The consequence of `logging.getLogger('YOURPLUGINNAME')`  is that, as it's not in the same namespace as ` sopel.*`, it won't use the sopel's logging handler, so it might prevent it from logging into the console/the log file.

For that, a plugin's `setup` function will have to configure manually a log handler for the plugin. This is not ideal, and I need to craft some helper for that.

Reply:
I'd guess that was ultimately the idea behind Sopel's `get_logger` function, even if the implementation was a mess: Give plugins a way to get a logger that always outputs to where the bot's logs should go, regardless of package naming, because the function handled figuring out the namespace.

I hope you're still thinking to keep plugin logs in the same place(s) as Sopel's main logs? Otherwise the log directory becomes pretty messy if there are potentially up to two files for _each plugin_ in addition to Sopel itself.

Reply:
I just realized two things:

* `sopel.logger`  is undocumented, which means that `sopel.logger.get_logger` is undocumented; we might consider changing its behavior without a warning
* we could log everything into one logging file, AND into a dedicated file for every plugins, i.e. log twice everything that come from a plugin in, say, `default.plugins.log`

My idea with `get_logger` would be to create a new function `get_plugin_logger`, that would return a logger for `sopel.externals.<plugin-name>`, and `get_logger` would call that function. Internally, we can configure Sopel logging with:

* log everything under `sopel.*` into `default.sopel.log`
* log everything under `sopel.externals.*`  into `default.plugins.log`  AND `default.sopel.log`

And even better than that: we can have a ListAttribute to deactivate logging for a list of plugins, something like `logging_plugin_disable`, and all I'd have to do is to configure `sopel.externals.<plugin_name>` with "don't log anything".

That could be cool.

Another option would be to allow a sub-section of the configuration for logging, where each key is a plugin-name, and each value is a log level, something likes this:

```
[core]
# ...
[core:logging]
plugin_name_1 = INFO
plugin_name_2 = DEBUG
plugin_name_3 = ERROR
```